sql/plan/case.go

```diff
@@ -0,0 +1,174 @@
+func (c *CaseStatement) Schema() sql.Schema {
+	return c.IfElse.Schema()
+}
+
+// Children implements the interface sql.Node.
+func (c *CaseStatement) Children() []sql.Node {
+	return c.IfElse.Children()
+}
+
+// WithChildren implements the interface sql.Node.
+func (c *CaseStatement) WithChildren(children ...sql.Node) (sql.Node, error) {
+	newIfElseNode, err := c.IfElse.WithChildren(children...)
+	if err != nil {
+		return nil, err
+	}
+	newIfElse, ok := newIfElseNode.(*IfElseBlock)
+	if !ok {
+		return nil, fmt.Errorf("%T: expected child %T but got %T", c, c.IfElse, newIfElseNode)
+	}
+
+	return &CaseStatement{
+		Expr:   c.Expr,
+		IfElse: newIfElse,
+	}, nil
+}
+
+// Expressions implements the interface sql.Node.
+func (c *CaseStatement) Expressions() []sql.Expression {
+	return []sql.Expression{c.Expr}
+}
+
+// WithExpressions implements the interface sql.Node.
+func (c *CaseStatement) WithExpressions(exprs ...sql.Expression) (sql.Node, error) {
+	if len(exprs) != 1 {
+		return nil, sql.ErrInvalidChildrenNumber.New(c, len(exprs), 1)
+	}
+
+	return &CaseStatement{
+		Expr:   exprs[0],
+		IfElse: c.IfElse,
+	}, nil
+}
+
+// CheckPrivileges implements the interface sql.Node.
+func (c *CaseStatement) CheckPrivileges(ctx *sql.Context, opChecker sql.PrivilegedOperationChecker) bool {
+	return c.IfElse.CheckPrivileges(ctx, opChecker)
+}
+
+// RowIter implements the interface sql.Node.
+func (c *CaseStatement) RowIter(ctx *sql.Context, row sql.Row) (sql.RowIter, error) {
+	var branchIter sql.RowIter
+
+	caseValue, err := c.Expr.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, ifConditional := range c.IfElse.IfConditionals {
+		whenValue, err := ifConditional.Condition.Eval(ctx, row)
+		if err != nil {
+			return nil, err
+		}
+		comparison, err := c.Expr.Type().Compare(caseValue, whenValue)
+		if err != nil {
+			return nil, err
+		}
+		if comparison != 0 {
+			continue
+		}
+
+		branchIter, err = ifConditional.RowIter(ctx, row)
+		if err != nil {
+			return nil, err
+		}
+		// If the branchIter is already a block iter, then we don't need to construct our own, as its contained
+		// node and schema will be a better representation of the iterated rows.
+		if blockRowIter, ok := branchIter.(BlockRowIter); ok {
+			return blockRowIter, nil
+		}
+		return &ifElseIter{
+			branchIter: branchIter,
+			sch:        ifConditional.Body.Schema(),
+			branchNode: ifConditional.Body,
+		}, nil
+	}
+
+	// All conditions failed so we run the else
+	branchIter, err = c.IfElse.Else.RowIter(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+	// If the branchIter is already a block iter, then we don't need to construct our own, as its contained
+	// node and schema will be a better representation of the iterated rows.
+	if blockRowIter, ok := branchIter.(BlockRowIter); ok {
+		return blockRowIter, nil
+	}
+	return &ifElseIter{
+		branchIter: branchIter,
+		sch:        c.IfElse.Else.Schema(),
+		branchNode: c.IfElse.Else,
+	}, nil
```

Comment:
This seems like it'd be another helpful helper function to extract since the code is needed twice in this function.